// @aztec/noir-contracts.js/src/AvmGadgetsTest.ts

/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractStorageLayout,
  type ContractNotes,
  decodeFromAbi,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  loadContractArtifact,
  loadContractArtifactForPublic,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type Wallet,
  type U128Like,
  type WrappedFieldLike,
} from '@aztec/aztec.js';
import AvmGadgetsTestContractArtifactJson from '../artifacts/avm_gadgets_test_contract-AvmGadgetsTest.json' assert { type: 'json' };
export const AvmGadgetsTestContractArtifact = loadContractArtifact(AvmGadgetsTestContractArtifactJson as NoirCompiledContract);



/**
 * Type-safe interface for contract AvmGadgetsTest;
 */
export class AvmGadgetsTestContract extends ContractBase {
  
  private constructor(
    instance: ContractInstanceWithAddress,
    wallet: Wallet,
  ) {
    super(instance, AvmGadgetsTestContractArtifact, wallet);
  }
  

  
  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(
    address: AztecAddress,
    wallet: Wallet,
  ) {
    return Contract.at(address, AvmGadgetsTestContract.artifact, wallet) as Promise<AvmGadgetsTestContract>;
  }

  
  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, ) {
    return new DeployMethod<AvmGadgetsTestContract>(PublicKeys.default(), wallet, AvmGadgetsTestContractArtifact, AvmGadgetsTestContract.at, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, ) {
    return new DeployMethod<AvmGadgetsTestContract>(publicKeys, wallet, AvmGadgetsTestContractArtifact, AvmGadgetsTestContract.at, Array.from(arguments).slice(2));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof AvmGadgetsTestContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<AvmGadgetsTestContract['methods'][M]>
  ) {
    return new DeployMethod<AvmGadgetsTestContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      AvmGadgetsTestContractArtifact,
      AvmGadgetsTestContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }
  

  
  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return AvmGadgetsTestContractArtifact;
  }

  /**
   * Returns this contract's artifact with public bytecode.
   */
  public static get artifactForPublic(): ContractArtifact {
    return loadContractArtifactForPublic(AvmGadgetsTestContractArtifactJson as NoirCompiledContract);
  }
  

  

  

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    
    /** keccak_f1600(data: array) */
    keccak_f1600: ((data: (bigint | number)[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** keccak_hash(data: array) */
    keccak_hash: ((data: (bigint | number)[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** pedersen_hash(data: array) */
    pedersen_hash: ((data: FieldLike[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** pedersen_hash_with_index(data: array) */
    pedersen_hash_with_index: ((data: FieldLike[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** poseidon2_hash(data: array) */
    poseidon2_hash: ((data: FieldLike[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sha256_hash_10(data: array) */
    sha256_hash_10: ((data: (bigint | number)[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sha256_hash_100(data: array) */
    sha256_hash_100: ((data: (bigint | number)[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sha256_hash_20(data: array) */
    sha256_hash_20: ((data: (bigint | number)[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sha256_hash_2048(data: array) */
    sha256_hash_2048: ((data: (bigint | number)[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sha256_hash_255(data: array) */
    sha256_hash_255: ((data: (bigint | number)[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sha256_hash_256(data: array) */
    sha256_hash_256: ((data: (bigint | number)[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sha256_hash_30(data: array) */
    sha256_hash_30: ((data: (bigint | number)[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sha256_hash_40(data: array) */
    sha256_hash_40: ((data: (bigint | number)[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sha256_hash_50(data: array) */
    sha256_hash_50: ((data: (bigint | number)[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sha256_hash_511(data: array) */
    sha256_hash_511: ((data: (bigint | number)[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sha256_hash_512(data: array) */
    sha256_hash_512: ((data: (bigint | number)[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sha256_hash_60(data: array) */
    sha256_hash_60: ((data: (bigint | number)[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sha256_hash_70(data: array) */
    sha256_hash_70: ((data: (bigint | number)[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sha256_hash_80(data: array) */
    sha256_hash_80: ((data: (bigint | number)[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sha256_hash_90(data: array) */
    sha256_hash_90: ((data: (bigint | number)[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sync_notes() */
    sync_notes: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };

  
}
