// @aztec/noir-contracts.js/src/ContractInstanceDeployer.ts

/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractStorageLayout,
  type ContractNotes,
  decodeFromAbi,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  loadContractArtifact,
  loadContractArtifactForPublic,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type Wallet,
  type U128Like,
  type WrappedFieldLike,
} from '@aztec/aztec.js';
import ContractInstanceDeployerContractArtifactJson from '../artifacts/contract_instance_deployer_contract-ContractInstanceDeployer.json' assert { type: 'json' };
export const ContractInstanceDeployerContractArtifact = loadContractArtifact(ContractInstanceDeployerContractArtifactJson as NoirCompiledContract);


      export type ContractInstanceUpdated = {
        DEPLOYER_CONTRACT_INSTANCE_UPDATED_MAGIC_VALUE: FieldLike
address: AztecAddressLike
prev_contract_class_id: WrappedFieldLike
new_contract_class_id: WrappedFieldLike
block_of_change: (bigint | number)
      }
    

      export type ContractInstanceDeployed = {
        DEPLOYER_CONTRACT_INSTANCE_DEPLOYED_MAGIC_VALUE: FieldLike
address: AztecAddressLike
version: (bigint | number)
salt: FieldLike
contract_class_id: WrappedFieldLike
initialization_hash: FieldLike
public_keys: { npk_m: { inner: { x: FieldLike, y: FieldLike, is_infinite: boolean } }, ivpk_m: { inner: { x: FieldLike, y: FieldLike, is_infinite: boolean } }, ovpk_m: { inner: { x: FieldLike, y: FieldLike, is_infinite: boolean } }, tpk_m: { inner: { x: FieldLike, y: FieldLike, is_infinite: boolean } } }
deployer: AztecAddressLike
      }
    

/**
 * Type-safe interface for contract ContractInstanceDeployer;
 */
export class ContractInstanceDeployerContract extends ContractBase {
  
  private constructor(
    instance: ContractInstanceWithAddress,
    wallet: Wallet,
  ) {
    super(instance, ContractInstanceDeployerContractArtifact, wallet);
  }
  

  
  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(
    address: AztecAddress,
    wallet: Wallet,
  ) {
    return Contract.at(address, ContractInstanceDeployerContract.artifact, wallet) as Promise<ContractInstanceDeployerContract>;
  }

  
  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, ) {
    return new DeployMethod<ContractInstanceDeployerContract>(PublicKeys.default(), wallet, ContractInstanceDeployerContractArtifact, ContractInstanceDeployerContract.at, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, ) {
    return new DeployMethod<ContractInstanceDeployerContract>(publicKeys, wallet, ContractInstanceDeployerContractArtifact, ContractInstanceDeployerContract.at, Array.from(arguments).slice(2));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof ContractInstanceDeployerContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<ContractInstanceDeployerContract['methods'][M]>
  ) {
    return new DeployMethod<ContractInstanceDeployerContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      ContractInstanceDeployerContractArtifact,
      ContractInstanceDeployerContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }
  

  
  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return ContractInstanceDeployerContractArtifact;
  }

  /**
   * Returns this contract's artifact with public bytecode.
   */
  public static get artifactForPublic(): ContractArtifact {
    return loadContractArtifactForPublic(ContractInstanceDeployerContractArtifactJson as NoirCompiledContract);
  }
  

  public static get storage(): ContractStorageLayout<'updated_class_ids'> {
      return {
        updated_class_ids: {
      slot: new Fr(1n),
    }
      } as ContractStorageLayout<'updated_class_ids'>;
    }
    

  

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    
    /** deploy(salt: field, contract_class_id: struct, initialization_hash: field, public_keys: struct, universal_deploy: boolean) */
    deploy: ((salt: FieldLike, contract_class_id: WrappedFieldLike, initialization_hash: FieldLike, public_keys: { npk_m: { inner: { x: FieldLike, y: FieldLike, is_infinite: boolean } }, ivpk_m: { inner: { x: FieldLike, y: FieldLike, is_infinite: boolean } }, ovpk_m: { inner: { x: FieldLike, y: FieldLike, is_infinite: boolean } }, tpk_m: { inner: { x: FieldLike, y: FieldLike, is_infinite: boolean } } }, universal_deploy: boolean) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_update_delay() */
    get_update_delay: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** set_update_delay(new_update_delay: integer) */
    set_update_delay: ((new_update_delay: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sync_notes() */
    sync_notes: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** update(new_contract_class_id: struct) */
    update: ((new_contract_class_id: WrappedFieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };

  
    public static get events(): { ContractInstanceUpdated: {abiType: AbiType, eventSelector: EventSelector, fieldNames: string[] }, ContractInstanceDeployed: {abiType: AbiType, eventSelector: EventSelector, fieldNames: string[] } } {
    return {
      ContractInstanceUpdated: {
        abiType: {
    "kind": "struct",
    "fields": [
        {
            "name": "DEPLOYER_CONTRACT_INSTANCE_UPDATED_MAGIC_VALUE",
            "type": {
                "kind": "field"
            }
        },
        {
            "name": "address",
            "type": {
                "kind": "struct",
                "fields": [
                    {
                        "name": "inner",
                        "type": {
                            "kind": "field"
                        }
                    }
                ],
                "path": "aztec::protocol_types::address::aztec_address::AztecAddress"
            }
        },
        {
            "name": "prev_contract_class_id",
            "type": {
                "kind": "struct",
                "fields": [
                    {
                        "name": "inner",
                        "type": {
                            "kind": "field"
                        }
                    }
                ],
                "path": "aztec::protocol_types::contract_class_id::ContractClassId"
            }
        },
        {
            "name": "new_contract_class_id",
            "type": {
                "kind": "struct",
                "fields": [
                    {
                        "name": "inner",
                        "type": {
                            "kind": "field"
                        }
                    }
                ],
                "path": "aztec::protocol_types::contract_class_id::ContractClassId"
            }
        },
        {
            "name": "block_of_change",
            "type": {
                "kind": "integer",
                "sign": "unsigned",
                "width": 32
            }
        }
    ],
    "path": "ContractInstanceDeployer::ContractInstanceUpdated"
},
        eventSelector: EventSelector.fromString("0x862aa094"),
        fieldNames: ["DEPLOYER_CONTRACT_INSTANCE_UPDATED_MAGIC_VALUE","address","prev_contract_class_id","new_contract_class_id","block_of_change"],
      },
ContractInstanceDeployed: {
        abiType: {
    "kind": "struct",
    "fields": [
        {
            "name": "DEPLOYER_CONTRACT_INSTANCE_DEPLOYED_MAGIC_VALUE",
            "type": {
                "kind": "field"
            }
        },
        {
            "name": "address",
            "type": {
                "kind": "struct",
                "fields": [
                    {
                        "name": "inner",
                        "type": {
                            "kind": "field"
                        }
                    }
                ],
                "path": "aztec::protocol_types::address::aztec_address::AztecAddress"
            }
        },
        {
            "name": "version",
            "type": {
                "kind": "integer",
                "sign": "unsigned",
                "width": 8
            }
        },
        {
            "name": "salt",
            "type": {
                "kind": "field"
            }
        },
        {
            "name": "contract_class_id",
            "type": {
                "kind": "struct",
                "fields": [
                    {
                        "name": "inner",
                        "type": {
                            "kind": "field"
                        }
                    }
                ],
                "path": "aztec::protocol_types::contract_class_id::ContractClassId"
            }
        },
        {
            "name": "initialization_hash",
            "type": {
                "kind": "field"
            }
        },
        {
            "name": "public_keys",
            "type": {
                "kind": "struct",
                "fields": [
                    {
                        "name": "npk_m",
                        "type": {
                            "kind": "struct",
                            "fields": [
                                {
                                    "name": "inner",
                                    "type": {
                                        "kind": "struct",
                                        "fields": [
                                            {
                                                "name": "x",
                                                "type": {
                                                    "kind": "field"
                                                }
                                            },
                                            {
                                                "name": "y",
                                                "type": {
                                                    "kind": "field"
                                                }
                                            },
                                            {
                                                "name": "is_infinite",
                                                "type": {
                                                    "kind": "boolean"
                                                }
                                            }
                                        ],
                                        "path": "std::embedded_curve_ops::EmbeddedCurvePoint"
                                    }
                                }
                            ],
                            "path": "aztec::protocol_types::public_keys::NpkM"
                        }
                    },
                    {
                        "name": "ivpk_m",
                        "type": {
                            "kind": "struct",
                            "fields": [
                                {
                                    "name": "inner",
                                    "type": {
                                        "kind": "struct",
                                        "fields": [
                                            {
                                                "name": "x",
                                                "type": {
                                                    "kind": "field"
                                                }
                                            },
                                            {
                                                "name": "y",
                                                "type": {
                                                    "kind": "field"
                                                }
                                            },
                                            {
                                                "name": "is_infinite",
                                                "type": {
                                                    "kind": "boolean"
                                                }
                                            }
                                        ],
                                        "path": "std::embedded_curve_ops::EmbeddedCurvePoint"
                                    }
                                }
                            ],
                            "path": "aztec::protocol_types::public_keys::IvpkM"
                        }
                    },
                    {
                        "name": "ovpk_m",
                        "type": {
                            "kind": "struct",
                            "fields": [
                                {
                                    "name": "inner",
                                    "type": {
                                        "kind": "struct",
                                        "fields": [
                                            {
                                                "name": "x",
                                                "type": {
                                                    "kind": "field"
                                                }
                                            },
                                            {
                                                "name": "y",
                                                "type": {
                                                    "kind": "field"
                                                }
                                            },
                                            {
                                                "name": "is_infinite",
                                                "type": {
                                                    "kind": "boolean"
                                                }
                                            }
                                        ],
                                        "path": "std::embedded_curve_ops::EmbeddedCurvePoint"
                                    }
                                }
                            ],
                            "path": "aztec::protocol_types::public_keys::OvpkM"
                        }
                    },
                    {
                        "name": "tpk_m",
                        "type": {
                            "kind": "struct",
                            "fields": [
                                {
                                    "name": "inner",
                                    "type": {
                                        "kind": "struct",
                                        "fields": [
                                            {
                                                "name": "x",
                                                "type": {
                                                    "kind": "field"
                                                }
                                            },
                                            {
                                                "name": "y",
                                                "type": {
                                                    "kind": "field"
                                                }
                                            },
                                            {
                                                "name": "is_infinite",
                                                "type": {
                                                    "kind": "boolean"
                                                }
                                            }
                                        ],
                                        "path": "std::embedded_curve_ops::EmbeddedCurvePoint"
                                    }
                                }
                            ],
                            "path": "aztec::protocol_types::public_keys::TpkM"
                        }
                    }
                ],
                "path": "aztec::protocol_types::public_keys::PublicKeys"
            }
        },
        {
            "name": "deployer",
            "type": {
                "kind": "struct",
                "fields": [
                    {
                        "name": "inner",
                        "type": {
                            "kind": "field"
                        }
                    }
                ],
                "path": "aztec::protocol_types::address::aztec_address::AztecAddress"
            }
        }
    ],
    "path": "ContractInstanceDeployer::ContractInstanceDeployed"
},
        eventSelector: EventSelector.fromString("0x7c99f494"),
        fieldNames: ["DEPLOYER_CONTRACT_INSTANCE_DEPLOYED_MAGIC_VALUE","address","version","salt","contract_class_id","initialization_hash","public_keys","deployer"],
      }
    };
  }
  
}
