// @aztec/noir-contracts.js/src/FPC.ts

/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractStorageLayout,
  type ContractNotes,
  decodeFromAbi,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  loadContractArtifact,
  loadContractArtifactForPublic,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type Wallet,
  type U128Like,
  type WrappedFieldLike,
} from '@aztec/aztec.js';
import FPCContractArtifactJson from '../artifacts/fpc_contract-FPC.json' assert { type: 'json' };
export const FPCContractArtifact = loadContractArtifact(FPCContractArtifactJson as NoirCompiledContract);



/**
 * Type-safe interface for contract FPC;
 */
export class FPCContract extends ContractBase {
  
  private constructor(
    instance: ContractInstanceWithAddress,
    wallet: Wallet,
  ) {
    super(instance, FPCContractArtifact, wallet);
  }
  

  
  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(
    address: AztecAddress,
    wallet: Wallet,
  ) {
    return Contract.at(address, FPCContract.artifact, wallet) as Promise<FPCContract>;
  }

  
  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, accepted_asset: AztecAddressLike, admin: AztecAddressLike) {
    return new DeployMethod<FPCContract>(PublicKeys.default(), wallet, FPCContractArtifact, FPCContract.at, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, accepted_asset: AztecAddressLike, admin: AztecAddressLike) {
    return new DeployMethod<FPCContract>(publicKeys, wallet, FPCContractArtifact, FPCContract.at, Array.from(arguments).slice(2));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof FPCContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<FPCContract['methods'][M]>
  ) {
    return new DeployMethod<FPCContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      FPCContractArtifact,
      FPCContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }
  

  
  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return FPCContractArtifact;
  }

  /**
   * Returns this contract's artifact with public bytecode.
   */
  public static get artifactForPublic(): ContractArtifact {
    return loadContractArtifactForPublic(FPCContractArtifactJson as NoirCompiledContract);
  }
  

  public static get storage(): ContractStorageLayout<'config'> {
      return {
        config: {
      slot: new Fr(1n),
    }
      } as ContractStorageLayout<'config'>;
    }
    

  public static get notes(): ContractNotes<'UintNote'> {
    return {
      UintNote: {
          id: new NoteSelector(0),
        }
    } as ContractNotes<'UintNote'>;
  }
  

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    
    /** constructor(accepted_asset: struct, admin: struct) */
    constructor: ((accepted_asset: AztecAddressLike, admin: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** fee_entrypoint_private(max_fee: integer, nonce: field) */
    fee_entrypoint_private: ((max_fee: (bigint | number), nonce: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** fee_entrypoint_public(max_fee: integer, nonce: field) */
    fee_entrypoint_public: ((max_fee: (bigint | number), nonce: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_accepted_asset() */
    get_accepted_asset: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** pull_funds(to: struct) */
    pull_funds: ((to: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sync_notes() */
    sync_notes: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };

  
}
