// @aztec/noir-contracts.js/src/NoConstructor.ts

/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractStorageLayout,
  type ContractNotes,
  decodeFromAbi,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  loadContractArtifact,
  loadContractArtifactForPublic,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type Wallet,
  type U128Like,
  type WrappedFieldLike,
} from '@aztec/aztec.js';
import NoConstructorContractArtifactJson from '../artifacts/no_constructor_contract-NoConstructor.json' assert { type: 'json' };
export const NoConstructorContractArtifact = loadContractArtifact(NoConstructorContractArtifactJson as NoirCompiledContract);



/**
 * Type-safe interface for contract NoConstructor;
 */
export class NoConstructorContract extends ContractBase {
  
  private constructor(
    instance: ContractInstanceWithAddress,
    wallet: Wallet,
  ) {
    super(instance, NoConstructorContractArtifact, wallet);
  }
  

  
  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(
    address: AztecAddress,
    wallet: Wallet,
  ) {
    return Contract.at(address, NoConstructorContract.artifact, wallet) as Promise<NoConstructorContract>;
  }

  
  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, ) {
    return new DeployMethod<NoConstructorContract>(PublicKeys.default(), wallet, NoConstructorContractArtifact, NoConstructorContract.at, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, ) {
    return new DeployMethod<NoConstructorContract>(publicKeys, wallet, NoConstructorContractArtifact, NoConstructorContract.at, Array.from(arguments).slice(2));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof NoConstructorContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<NoConstructorContract['methods'][M]>
  ) {
    return new DeployMethod<NoConstructorContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      NoConstructorContractArtifact,
      NoConstructorContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }
  

  
  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return NoConstructorContractArtifact;
  }

  /**
   * Returns this contract's artifact with public bytecode.
   */
  public static get artifactForPublic(): ContractArtifact {
    return loadContractArtifactForPublic(NoConstructorContractArtifactJson as NoirCompiledContract);
  }
  

  public static get storage(): ContractStorageLayout<'private_mutable'> {
      return {
        private_mutable: {
      slot: new Fr(1n),
    }
      } as ContractStorageLayout<'private_mutable'>;
    }
    

  public static get notes(): ContractNotes<'ValueNote'> {
    return {
      ValueNote: {
          id: new NoteSelector(0),
        }
    } as ContractNotes<'ValueNote'>;
  }
  

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    
    /** emit_public(value: field) */
    emit_public: ((value: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** initialize_private_mutable(value: field) */
    initialize_private_mutable: ((value: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** is_private_mutable_initialized() */
    is_private_mutable_initialized: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sync_notes() */
    sync_notes: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };

  
}
