// @aztec/noir-contracts.js/src/StateVars.ts

/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractStorageLayout,
  type ContractNotes,
  decodeFromAbi,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  loadContractArtifact,
  loadContractArtifactForPublic,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type Wallet,
  type U128Like,
  type WrappedFieldLike,
} from '@aztec/aztec.js';
import StateVarsContractArtifactJson from '../artifacts/state_vars_contract-StateVars.json' assert { type: 'json' };
export const StateVarsContractArtifact = loadContractArtifact(StateVarsContractArtifactJson as NoirCompiledContract);



/**
 * Type-safe interface for contract StateVars;
 */
export class StateVarsContract extends ContractBase {
  
  private constructor(
    instance: ContractInstanceWithAddress,
    wallet: Wallet,
  ) {
    super(instance, StateVarsContractArtifact, wallet);
  }
  

  
  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(
    address: AztecAddress,
    wallet: Wallet,
  ) {
    return Contract.at(address, StateVarsContract.artifact, wallet) as Promise<StateVarsContract>;
  }

  
  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, ) {
    return new DeployMethod<StateVarsContract>(PublicKeys.default(), wallet, StateVarsContractArtifact, StateVarsContract.at, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, ) {
    return new DeployMethod<StateVarsContract>(publicKeys, wallet, StateVarsContractArtifact, StateVarsContract.at, Array.from(arguments).slice(2));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof StateVarsContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<StateVarsContract['methods'][M]>
  ) {
    return new DeployMethod<StateVarsContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      StateVarsContractArtifact,
      StateVarsContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }
  

  
  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return StateVarsContractArtifact;
  }

  /**
   * Returns this contract's artifact with public bytecode.
   */
  public static get artifactForPublic(): ContractArtifact {
    return loadContractArtifactForPublic(StateVarsContractArtifactJson as NoirCompiledContract);
  }
  

  

  public static get notes(): ContractNotes<'ValueNote'> {
    return {
      ValueNote: {
          id: new NoteSelector(0),
        }
    } as ContractNotes<'ValueNote'>;
  }
  

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    
    /** get_private_mutable() */
    get_private_mutable: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_public_immutable() */
    get_public_immutable: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_public_immutable_constrained_private() */
    get_public_immutable_constrained_private: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_public_immutable_constrained_private_indirect() */
    get_public_immutable_constrained_private_indirect: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_public_immutable_constrained_public() */
    get_public_immutable_constrained_public: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_public_immutable_constrained_public_indirect() */
    get_public_immutable_constrained_public_indirect: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_public_immutable_constrained_public_multiple() */
    get_public_immutable_constrained_public_multiple: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** increase_private_value() */
    increase_private_value: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** initialize_private(randomness: field, value: field) */
    initialize_private: ((randomness: FieldLike, value: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** initialize_private_immutable(randomness: field, value: field) */
    initialize_private_immutable: ((randomness: FieldLike, value: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** initialize_public_immutable(value: integer) */
    initialize_public_immutable: ((value: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** is_priv_imm_initialized() */
    is_priv_imm_initialized: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** is_private_mutable_initialized() */
    is_private_mutable_initialized: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** match_public_immutable(account: struct, value: integer) */
    match_public_immutable: ((account: AztecAddressLike, value: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sync_notes() */
    sync_notes: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** update_private_mutable(randomness: field, value: field) */
    update_private_mutable: ((randomness: FieldLike, value: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** view_private_immutable() */
    view_private_immutable: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };

  
}
