// @aztec/noir-contracts.js/src/Test.ts

/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractStorageLayout,
  type ContractNotes,
  decodeFromAbi,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  loadContractArtifact,
  loadContractArtifactForPublic,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type Wallet,
  type U128Like,
  type WrappedFieldLike,
} from '@aztec/aztec.js';
import TestContractArtifactJson from '../artifacts/test_contract-Test.json' assert { type: 'json' };
export const TestContractArtifact = loadContractArtifact(TestContractArtifactJson as NoirCompiledContract);


      export type ExampleEvent = {
        value0: FieldLike
value1: FieldLike
value2: FieldLike
value3: FieldLike
value4: FieldLike
      }
    

/**
 * Type-safe interface for contract Test;
 */
export class TestContract extends ContractBase {
  
  private constructor(
    instance: ContractInstanceWithAddress,
    wallet: Wallet,
  ) {
    super(instance, TestContractArtifact, wallet);
  }
  

  
  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(
    address: AztecAddress,
    wallet: Wallet,
  ) {
    return Contract.at(address, TestContract.artifact, wallet) as Promise<TestContract>;
  }

  
  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, ) {
    return new DeployMethod<TestContract>(PublicKeys.default(), wallet, TestContractArtifact, TestContract.at, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, ) {
    return new DeployMethod<TestContract>(publicKeys, wallet, TestContractArtifact, TestContract.at, Array.from(arguments).slice(2));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof TestContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<TestContract['methods'][M]>
  ) {
    return new DeployMethod<TestContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      TestContractArtifact,
      TestContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }
  

  
  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return TestContractArtifact;
  }

  /**
   * Returns this contract's artifact with public bytecode.
   */
  public static get artifactForPublic(): ContractArtifact {
    return loadContractArtifactForPublic(TestContractArtifactJson as NoirCompiledContract);
  }
  

  public static get storage(): ContractStorageLayout<'example_constant' | 'example_struct_in_shared_mutable' | 'example_set' | 'example_struct' | 'example_struct_in_public_immutable' | 'example_struct_in_map' | 'another_example_struct'> {
      return {
        example_constant: {
      slot: new Fr(1n),
    },
example_struct_in_shared_mutable: {
      slot: new Fr(2n),
    },
example_set: {
      slot: new Fr(14n),
    },
example_struct: {
      slot: new Fr(15n),
    },
example_struct_in_public_immutable: {
      slot: new Fr(20n),
    },
example_struct_in_map: {
      slot: new Fr(26n),
    },
another_example_struct: {
      slot: new Fr(27n),
    }
      } as ContractStorageLayout<'example_constant' | 'example_struct_in_shared_mutable' | 'example_set' | 'example_struct' | 'example_struct_in_public_immutable' | 'example_struct_in_map' | 'another_example_struct'>;
    }
    

  public static get notes(): ContractNotes<'AddressNote' | 'TestNote' | 'ValueNote'> {
    return {
      AddressNote: {
          id: new NoteSelector(0),
        },
TestNote: {
          id: new NoteSelector(2),
        },
ValueNote: {
          id: new NoteSelector(1),
        }
    } as ContractNotes<'AddressNote' | 'TestNote' | 'ValueNote'>;
  }
  

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    
    /** assert_header_private(header_hash: field) */
    assert_header_private: ((header_hash: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** assert_private_global_vars(chain_id: field, version: field) */
    assert_private_global_vars: ((chain_id: FieldLike, version: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** assert_public_global_vars(chain_id: field, version: field, block_number: field, timestamp: integer, fee_per_da_gas: field, fee_per_l2_gas: field) */
    assert_public_global_vars: ((chain_id: FieldLike, version: FieldLike, block_number: FieldLike, timestamp: (bigint | number), fee_per_da_gas: FieldLike, fee_per_l2_gas: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** call_create_note(value: field, owner: struct, sender: struct, storage_slot: field) */
    call_create_note: ((value: FieldLike, owner: AztecAddressLike, sender: AztecAddressLike, storage_slot: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** call_destroy_note(storage_slot: field) */
    call_destroy_note: ((storage_slot: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** call_get_notes(storage_slot: field, active_or_nullified: boolean) */
    call_get_notes: ((storage_slot: FieldLike, active_or_nullified: boolean) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** call_get_notes_many(storage_slot: field, active_or_nullified: boolean) */
    call_get_notes_many: ((storage_slot: FieldLike, active_or_nullified: boolean) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** call_view_notes(storage_slot: field, active_or_nullified: boolean) */
    call_view_notes: ((storage_slot: FieldLike, active_or_nullified: boolean) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** call_view_notes_many(storage_slot: field, active_or_nullified: boolean) */
    call_view_notes_many: ((storage_slot: FieldLike, active_or_nullified: boolean) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** consume_message_from_arbitrary_sender_private(content: field, secret: field, sender: struct, message_leaf_index: field) */
    consume_message_from_arbitrary_sender_private: ((content: FieldLike, secret: FieldLike, sender: EthAddressLike, message_leaf_index: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** consume_message_from_arbitrary_sender_public(content: field, secret: field, sender: struct, message_leaf_index: field) */
    consume_message_from_arbitrary_sender_public: ((content: FieldLike, secret: FieldLike, sender: EthAddressLike, message_leaf_index: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** consume_mint_to_private_message(amount: integer, secret_for_L1_to_L2_message_consumption: field, portal_address: struct, message_leaf_index: field) */
    consume_mint_to_private_message: ((amount: (bigint | number), secret_for_L1_to_L2_message_consumption: FieldLike, portal_address: EthAddressLike, message_leaf_index: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** consume_mint_to_public_message(to: struct, amount: integer, secret: field, message_leaf_index: field, portal_address: struct) */
    consume_mint_to_public_message: ((to: AztecAddressLike, amount: (bigint | number), secret: FieldLike, message_leaf_index: FieldLike, portal_address: EthAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** create_l2_to_l1_message_arbitrary_recipient_private(content: field, recipient: struct) */
    create_l2_to_l1_message_arbitrary_recipient_private: ((content: FieldLike, recipient: EthAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** create_l2_to_l1_message_arbitrary_recipient_public(content: field, recipient: struct) */
    create_l2_to_l1_message_arbitrary_recipient_public: ((content: FieldLike, recipient: EthAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** create_l2_to_l1_message_public(amount: field, secret_hash: field, portal_address: struct) */
    create_l2_to_l1_message_public: ((amount: FieldLike, secret_hash: FieldLike, portal_address: EthAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** deliver_note(contract_address: struct, value: field, tx_hash: field, unique_note_hashes_in_tx: struct, first_nullifier_in_tx: field, recipient: struct) */
    deliver_note: ((contract_address: AztecAddressLike, value: FieldLike, tx_hash: FieldLike, unique_note_hashes_in_tx: { storage: FieldLike[], len: (bigint | number) }, first_nullifier_in_tx: FieldLike, recipient: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** deploy_contract(target: struct) */
    deploy_contract: ((target: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** emit_array_as_encrypted_log(fields: array, owner: struct, sender: struct, nest: boolean) */
    emit_array_as_encrypted_log: ((fields: FieldLike[], owner: AztecAddressLike, sender: AztecAddressLike, nest: boolean) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** emit_nullifier(nullifier: field) */
    emit_nullifier: ((nullifier: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** emit_nullifier_public(nullifier: field) */
    emit_nullifier_public: ((nullifier: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** emit_public(value: field) */
    emit_public: ((value: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_constant() */
    get_constant: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_master_incoming_viewing_public_key(address: struct) */
    get_master_incoming_viewing_public_key: ((address: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_ovsk_app(ovpk_m_hash: field) */
    get_ovsk_app: ((ovpk_m_hash: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_this_address() */
    get_this_address: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** initialize() */
    initialize: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** is_time_equal(time: integer) */
    is_time_equal: ((time: (bigint | number)) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** set_constant(value: field) */
    set_constant: ((value: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** set_tx_max_block_number(max_block_number: integer, enqueue_public_call: boolean) */
    set_tx_max_block_number: ((max_block_number: (bigint | number), enqueue_public_call: boolean) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sync_notes() */
    sync_notes: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** test_code_gen(a_field: field, a_bool: boolean, a_number: integer, an_array: array, a_struct: struct, a_deep_struct: struct) */
    test_code_gen: ((a_field: FieldLike, a_bool: boolean, a_number: (bigint | number), an_array: FieldLike[], a_struct: { amount: FieldLike, secret_hash: FieldLike }, a_deep_struct: { a_field: FieldLike, a_bool: boolean, a_note: { amount: FieldLike, secret_hash: FieldLike }, many_notes: { amount: FieldLike, secret_hash: FieldLike }[] }) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** test_note_inclusion(owner: struct, storage_slot: field) */
    test_note_inclusion: ((owner: AztecAddressLike, storage_slot: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** test_nullifier_key_freshness(address: struct, public_nullifying_key: struct) */
    test_nullifier_key_freshness: ((address: AztecAddressLike, public_nullifying_key: { x: FieldLike, y: FieldLike, is_infinite: boolean }) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** test_setting_fee_payer() */
    test_setting_fee_payer: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** test_setting_teardown() */
    test_setting_teardown: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };

  
    public static get events(): { ExampleEvent: {abiType: AbiType, eventSelector: EventSelector, fieldNames: string[] } } {
    return {
      ExampleEvent: {
        abiType: {
    "kind": "struct",
    "fields": [
        {
            "name": "value0",
            "type": {
                "kind": "field"
            }
        },
        {
            "name": "value1",
            "type": {
                "kind": "field"
            }
        },
        {
            "name": "value2",
            "type": {
                "kind": "field"
            }
        },
        {
            "name": "value3",
            "type": {
                "kind": "field"
            }
        },
        {
            "name": "value4",
            "type": {
                "kind": "field"
            }
        }
    ],
    "path": "Test::ExampleEvent"
},
        eventSelector: EventSelector.fromString("0xbb59ba6b"),
        fieldNames: ["value0","value1","value2","value3","value4"],
      }
    };
  }
  
}
