// @aztec/noir-contracts.js/src/Updated.ts

/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  type AbiType,
  AztecAddress,
  type AztecAddressLike,
  CompleteAddress,
  Contract,
  type ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  type ContractInstanceWithAddress,
  type ContractMethod,
  type ContractStorageLayout,
  type ContractNotes,
  decodeFromAbi,
  DeployMethod,
  EthAddress,
  type EthAddressLike,
  EventSelector,
  type FieldLike,
  Fr,
  type FunctionSelectorLike,
  loadContractArtifact,
  loadContractArtifactForPublic,
  type NoirCompiledContract,
  NoteSelector,
  Point,
  type PublicKey,
  PublicKeys,
  type Wallet,
  type U128Like,
  type WrappedFieldLike,
} from '@aztec/aztec.js';
import UpdatedContractArtifactJson from '../artifacts/updated_contract-Updated.json' assert { type: 'json' };
export const UpdatedContractArtifact = loadContractArtifact(UpdatedContractArtifactJson as NoirCompiledContract);



/**
 * Type-safe interface for contract Updated;
 */
export class UpdatedContract extends ContractBase {
  
  private constructor(
    instance: ContractInstanceWithAddress,
    wallet: Wallet,
  ) {
    super(instance, UpdatedContractArtifact, wallet);
  }
  

  
  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(
    address: AztecAddress,
    wallet: Wallet,
  ) {
    return Contract.at(address, UpdatedContract.artifact, wallet) as Promise<UpdatedContract>;
  }

  
  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet, ) {
    return new DeployMethod<UpdatedContract>(PublicKeys.default(), wallet, UpdatedContractArtifact, UpdatedContract.at, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public keys hash to derive the address.
   */
  public static deployWithPublicKeys(publicKeys: PublicKeys, wallet: Wallet, ) {
    return new DeployMethod<UpdatedContract>(publicKeys, wallet, UpdatedContractArtifact, UpdatedContract.at, Array.from(arguments).slice(2));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof UpdatedContract['methods']>(
    opts: { publicKeys?: PublicKeys; method?: M; wallet: Wallet },
    ...args: Parameters<UpdatedContract['methods'][M]>
  ) {
    return new DeployMethod<UpdatedContract>(
      opts.publicKeys ?? PublicKeys.default(),
      opts.wallet,
      UpdatedContractArtifact,
      UpdatedContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }
  

  
  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return UpdatedContractArtifact;
  }

  /**
   * Returns this contract's artifact with public bytecode.
   */
  public static get artifactForPublic(): ContractArtifact {
    return loadContractArtifactForPublic(UpdatedContractArtifactJson as NoirCompiledContract);
  }
  

  public static get storage(): ContractStorageLayout<'private_value' | 'public_value'> {
      return {
        private_value: {
      slot: new Fr(1n),
    },
public_value: {
      slot: new Fr(2n),
    }
      } as ContractStorageLayout<'private_value' | 'public_value'>;
    }
    

  public static get notes(): ContractNotes<'ValueNote'> {
    return {
      ValueNote: {
          id: new NoteSelector(0),
        }
    } as ContractNotes<'ValueNote'>;
  }
  

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public declare methods: {
    
    /** get_private_value() */
    get_private_value: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_public_value() */
    get_public_value: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_update_delay() */
    get_update_delay: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** set_private_value() */
    set_private_value: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** set_public_value() */
    set_public_value: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** sync_notes() */
    sync_notes: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };

  
}
